### ISSUE TITLE
Awaiting a send_request result hangs using the new async-await syntax.

### ISSUE THREAD

**Issue by DevQps** [hyperium/h2#390]
Hello, 

I apologize in advance if this not the right place to ask this question, but I thought I maybe ran into a bug. I posted this question on [StackOverflow](https://stackoverflow.com/questions/57307892/how-to-do-an-http2-request-with-h2-using-the-new-async-await-syntax-in-rust) as well. May this be solved then I will create an answer there as well.

The problem I run into is that I tried to convert the [H2 Akamai example](https://github.com/hyperium/h2/blob/master/examples/akamai.rs) into code using Rust's new async-await syntax.

I have been able to produce the following code, but it hangs on `let response = response.compat().await;` without me able to understand why.

```
#![feature(async_await)]

use tokio::net::TcpStream;
use std::sync::Arc;
use webpki::DNSNameRef;
use futures::compat::Future01CompatExt;
use futures::future::{FutureExt, TryFutureExt};
use h2::client;
use rustls::ClientConfig;
use tokio_rustls::ClientConfigExt;

use rustls::Session;

use std::net::ToSocketAddrs;
use hyper::{Method, Request};

pub fn setup_config() -> Arc<ClientConfig>
{
    std::sync::Arc::new({
        let mut c = rustls::ClientConfig::new();
        c.root_store
            .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
        c.alpn_protocols.push("h2".to_owned());
        c
    })
}

pub async fn worker()
{
    // Set the address to run our socket on.
    let address = "http2.akamai.com:443"
        .to_socket_addrs()
        .unwrap()
        .next()
        .unwrap();

    let config = setup_config();

    let dns_name = DNSNameRef::try_from_ascii_str("http2.akamai.com").unwrap();

    // Open a TCP connection.
    let tcp = TcpStream::connect(&address).compat().await.unwrap();
    ;

    let tls = config.connect_async(dns_name, tcp).compat().await.unwrap();

    let (_, session) = tls.get_ref();
    let negotiated_protocol = session.get_alpn_protocol();
    assert_eq!(Some("h2"), negotiated_protocol.as_ref().map(|x| &**x));

    let res = client::handshake(tls).compat().await;

    let (client, h2) = res.unwrap();

    println!("Test5");
    let request = Request::builder()
        .method(Method::GET)
        .uri("https://http2.akamai.com/")
        .body(())
        .unwrap();

    println!("Test6");

    let (response, x) = client.ready().compat().await.unwrap().send_request(request, true).unwrap();
    println!("Test7");

    let response = response.compat().await;
    println!("Test8");
}

fn main()
{
    // Call our `run_server` function, which returns a future.
    // As with every `async fn`, for `run_server` to do anything,
    // the returned future needs to be run. Additionally,
    // we need to convert the returned future from a futures 0.3 future into a
    // futures 0.1 future.
    let futures_03_future = worker();
    let futures_01_future = futures_03_future.unit_error().boxed().compat();

    // Finally, we can run the future to completion using the `run` function
    // provided by Hyper.
    tokio::run(futures_01_future);
}
```

Cargo.toml:
```
[dependencies]
# The latest version of the "futures" library, which has lots of utilities
# for writing async code. Enable the "compat" feature to include the
# functions for using futures 0.3 and async/await with the Hyper library,
# which use futures 0.1.
futures-preview = { version = "=0.3.0-alpha.16", features = ["compat"] }

# Hyper is an asynchronous HTTP library. We'll use it to power our HTTP
# server and to make HTTP requests.
hyper = "0.12.9"

# Tokio
tokio = "0.1.22"
h2 = "0.1.26"

# RustTLS
rustls = "0.12"
tokio-rustls = "0.5.0"
webpki = "0.18"
webpki-roots = "0.14"
```

Output:
```
Test5
Test6
Test7
```

I checked Wireshark as well, and the HTTP2 connection has been opened, but the request inside the connection is not being sent. But I still don't understand why.

Is this maybe some kind of bug? Or am I missing something?

Thanks in advance! I'd gladly make a PR for documentation as well if there is something that can be improved regarding this question.

**Comment by seanmonstar** [comment]
After the h2 handshake, you get a client and a connection back in a tuple. The connection is a `Future` you need to spawn onto an executor (if using tokio, then just `tokio::spawn`). It handles all the connection stuff, including the sending and receiving of requests and responses (and also flow control, pings, etc). Without spawning it, nothing on the connection will actually happen.

**Comment by DevQps** [comment]
Thanks a ton for your information! In the end I was able to get it to work with your hint!. I hope I'll be able to post an update version somewhere below here and on Reddit and Stackoverflow as well. (I also had some compatability issues with the futures package, but I solved that by aliasing futures 0.1 to a different name inside Cargo.toml)

Thanks again!

**Comment by DevQps** [comment]
@seanmonstar Again thanks for your help. While trying to read the responses contents I walked into another problem where 

```
let response = response.compat().await.unwrap();
let (_, body) = response.into_parts();
body.for_each(|chunk| {
            println!("RX: {:?}", chunk);
            Ok(())
        }).compat().await
```

receives a partial response and then hangs. I will create a new issue for this and close this one.
